estransport/connection.go

```diff
@@ -0,0 +1,280 @@
+//
+func (cp *roundRobinConnectionPool) Remove(c *Connection) error {
+	cp.Lock()
+	defer cp.Unlock()
+
+	c.Lock()
+
+	if c.Dead {
+		if cp.debugLogger != nil {
+			cp.debugLogger.Logf("Already removed %s\n", c.URL)
+		}
+		c.Unlock()
+		return nil
+	}
+
+	if cp.debugLogger != nil {
+		cp.debugLogger.Logf("Removing %s...\n", c.URL)
+	}
+	c.markAsDead()
+	cp.scheduleResurrect(c)
+	c.Unlock()
+
+	// Push item to dead list and sort slice by number of failures
+	cp.dead = append(cp.dead, c)
+	sort.Slice(cp.dead, func(i, j int) bool {
+		c1 := cp.dead[i]
+		c2 := cp.dead[j]
+		c1.Lock()
+		c2.Lock()
+		defer c1.Unlock()
+		defer c2.Unlock()
+
+		res := c1.Failures > c2.Failures
+		return res
+	})
+
+	// Check if connection exists in the list. Return nil if it doesn't,
+	// because another goroutine might have already removed it.
+	index := -1
+	for i, conn := range cp.live {
+		if conn == c {
+			index = i
+		}
+	}
+	if index < 0 {
+		return nil
+	}
+
+	// Remove item; https://github.com/golang/go/wiki/SliceTricks
+	copy(cp.live[index:], cp.live[index+1:])
+	cp.live = cp.live[:len(cp.live)-1]
+
+	if cp.enableMetrics {
+		cp.metrics.Lock()
+		cp.metrics.dead = cp.dead
+		cp.metrics.live = cp.live
+		cp.metrics.Unlock()
+	}
+
+	return nil
+}
+
+// Resurrect removes the connection from the dead list and adds it to the pool.
+//
+// TODO(karmi): Add a pluggable strategy as argument, eg. "optimistic", "ping".
+//
+func (cp *roundRobinConnectionPool) Resurrect(c *Connection) error {
+	cp.Lock()
+	defer cp.Unlock()
+
+	c.Lock()
+	defer c.Unlock()
+
+	if !c.Dead {
+		if cp.debugLogger != nil {
+			cp.debugLogger.Logf("Already resurrected %s\n", c.URL)
+		}
+		return nil
+	}
+
+	if cp.debugLogger != nil {
+		cp.debugLogger.Logf("Resurrecting %s\n", c.URL)
+	}
+
+	c.markAsLive()
+
+	cp.live = append(cp.live, c)
+	index := -1
+	for i, conn := range cp.dead {
+		if conn == c {
+			index = i
+		}
+	}
+	if index >= 0 {
+		// Remove item; https://github.com/golang/go/wiki/SliceTricks
+		copy(cp.dead[index:], cp.dead[index+1:])
+		cp.dead = cp.dead[:len(cp.dead)-1]
+	}
+
+	if cp.enableMetrics {
+		cp.metrics.Lock()
+		cp.metrics.dead = cp.dead
+		cp.metrics.live = cp.live
+		cp.metrics.Unlock()
+	}
+
+	return nil
+}
+
+// scheduleResurrect schedules the connection to be resurrected.
+//
+func (cp *roundRobinConnectionPool) scheduleResurrect(c *Connection) {
+	factor := math.Min(float64(c.Failures-1), float64(defaultResurrectTimeoutFactorCutoff))
+	timeout := time.Duration(defaultResurrectTimeoutInitial.Seconds() * math.Exp2(factor) * float64(time.Second))
+	if cp.debugLogger != nil {
+		cp.debugLogger.Logf("Resurrect %s (failures=%d, factor=%1.1f, timeout=%s) in %s\n", c.URL, c.Failures, factor, timeout, c.DeadSince.Add(timeout).Sub(time.Now().UTC()).Truncate(time.Second))
+	}
+
+	time.AfterFunc(timeout, func() { cp.Resurrect(c) })
+}
+
+// markAsDead marks the connection as dead.
+//
+func (c *Connection) markAsDead() {
```

Comment:
We might consider to export these methods given that users should be allowed to implement custom pools.